compiler-plugin/src/test/kotlin/com/google/devtools/ksp/test/AbstractKotlinKSPTest.kt

```diff
@@ -18,49 +18,69 @@
 
 package com.google.devtools.ksp.test
 
+import com.google.devtools.ksp.KotlinSymbolProcessingExtension
+import com.google.devtools.ksp.KspOptions
+import com.google.devtools.ksp.processing.impl.MessageCollectorBasedKSPLogger
+import com.google.devtools.ksp.processor.AbstractTestProcessor
 import junit.framework.TestCase
 import org.jetbrains.kotlin.cli.common.messages.MessageRenderer
 import org.jetbrains.kotlin.cli.common.messages.PrintingMessageCollector
+import org.jetbrains.kotlin.cli.jvm.compiler.EnvironmentConfigFiles
+import org.jetbrains.kotlin.cli.jvm.compiler.KotlinCoreEnvironment
+import org.jetbrains.kotlin.codegen.ClassBuilderFactories
 import org.jetbrains.kotlin.codegen.CodegenTestCase
 import org.jetbrains.kotlin.codegen.GenerationUtils
-import com.google.devtools.ksp.KotlinSymbolProcessingExtension
-import com.google.devtools.ksp.KspOptions
-import com.google.devtools.ksp.processing.impl.MessageCollectorBasedKSPLogger
-import com.google.devtools.ksp.processor.AbstractTestProcessor
 import org.jetbrains.kotlin.resolve.jvm.extensions.AnalysisHandlerExtension
-import org.jetbrains.kotlin.test.ConfigurationKind
+import org.jetbrains.kotlin.test.*
 import java.io.File
-import org.jetbrains.kotlin.test.TestJdkKind
 
-abstract class AbstractKotlinKSPTest : CodegenTestCase() {
+abstract class AbstractKotlinKSPTest : KotlinBaseTest<AbstractKotlinKSPTest.KspTestFile>() {
     companion object {
         const val TEST_PROCESSOR = "// TEST PROCESSOR:"
         val EXPECTED_RESULTS = "// EXPECTED:"
     }
+    private val testTmpDir by lazy {
+        KotlinTestUtils.tmpDir("test")
+    }
 
-    override fun doMultiFileTest(wholeFile: File, files: List<TestFile>) {
-        val javaFiles = listOfNotNull(writeJavaFiles(files))
-        createEnvironmentWithMockJdkAndIdeaAnnotations(ConfigurationKind.NO_KOTLIN_REFLECT, emptyList(), TestJdkKind.FULL_JDK_9, *(javaFiles.toTypedArray()))
+    override fun doMultiFileTest(wholeFile: File, files: List<KspTestFile>) {
+        // get the main module where KSP tests will be run. If there is no module declared in the test input, we'll
+        // create one that will contain all test files
+        val mainModule: TestModule = files.findLast {
+            it.testModule != null
+        }?.testModule ?: TestModule("main", emptyList(), emptyList())
+
+        // group each test file with its module
+        val filesByModule = groupFilesByModule(mainModule, files)
+
+        // now compile each sub module, we'll compile the main module last.
```

Comment:
Just curious, how is the module dependency affecting compiling order? Do you have to write modules in the topological order? If so, it's fine to have such a requirement but worth noting in the comment.

Reply:
yes. I could add a resolution to avoid that requirement but didn't think it is necessary and also creates a bit of ambiguity on where to apply KSP (if there are multiple leaf modules).

Reply:
added a comment to clarify this.